File: crates/eval-runner/src/state_query.rs
//! 12 state-query types (spec §4a). All execute via HTTP GET on the test server.
use crate::test_server::TestServer;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Deserialize)]
pub struct StateQueryInput {
    pub r#type: String,
    pub params: Value,
    pub expected: Value,
}

#[derive(Debug, Clone, Serialize)]
pub struct StateQueryResult {
    pub r#type: String,
    pub params: Value,
    pub expected: Value,
    pub actual: Value,
    pub passed: bool,
}

pub async fn run_query(q: &StateQueryInput, server: &TestServer) -> anyhow::Result<StateQueryResult> {
    let actual = match q.r#type.as_str() {
        "device_exists" => exists_at(server, &format!("/devices/{}", sid(&q.params, "id")?)).await?,
        "rule_exists" => exists_at(server, &format!("/rules/{}", sid(&q.params, "id")?)).await?,
        "agent_exists" => exists_at(server, &format!("/agents/{}", sid(&q.params, "id")?)).await?,
        "transform_exists" => {
            exists_at(server, &format!("/automations/{}", sid(&q.params, "id")?)).await?
        }
        "dashboard_exists" => {
            exists_at(server, &format!("/dashboards/{}", sid(&q.params, "id")?)).await?
        }
        "channel_exists" => {
            exists_at(server, &format!("/messages/channels/{}", sid(&q.params, "name")?)).await?
        }
        "rule_enabled" => {
            field_at(server, &format!("/rules/{}", sid(&q.params, "id")?), "enabled").await?
        }
        "agent_status" => {
            field_at(server, &format!("/agents/{}", sid(&q.params, "id")?), "status").await?
        }
        "push_enabled" => {
            field_at(server, &format!("/data-push/{}", sid(&q.params, "id")?), "enabled").await?
        }
        "device_count" => count_at(server, "/devices").await?,
        "dashboard_component_count" => {
            let v = get_json(server, &format!("/dashboards/{}", sid(&q.params, "id")?)).await?;
            Value::from(
                v.get("components")
                    .and_then(|c| c.as_array())
                    .map(|a| a.len())
                    .unwrap_or(0),
            )
        }
        "message_count" => count_at(server, "/messages").await?,
        other => anyhow::bail!("unknown state_query type: {}", other),
    };
    let passed = actual == q.expected;
    Ok(StateQueryResult {
        r#type: q.r#type.clone(),
        params: q.params.clone(),
        expected: q.expected.clone(),
        actual,
        passed,
    })
}

pub fn sid(params: &Value, key: &str) -> anyhow::Result<String> {
    Ok(params
        .get(key)
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("missing param {}", key))?
        .to_string())
}

async fn exists_at(server: &TestServer, path: &str) -> anyhow::Result<Value> {
    let resp = server.http_get(path).await?;
    Ok(Value::Bool(resp.status().as_u16() == 200))
}

pub async fn get_json(server: &TestServer, path: &str) -> anyhow::Result<Value> {
    let resp = server.http_get(path).await?;
    let body: Value = resp.json().await?;
    // Many NeoMind routes return {success, data} envelope.
    Ok(body.get("data").cloned().unwrap_or(body))
}

async fn field_at(server: &TestServer, path: &str, field: &str) -> anyhow::Result<Value> {
    let v = get_json(server, path).await?;
    Ok(v.get(field).cloned().unwrap_or(Value::Null))
}

async fn count_at(server: &TestServer, path: &str) -> anyhow::Result<Value> {
    let v = get_json(server, path).await?;
    let n = match v {
        Value::Array(a) => a.len(),
        _ => 0,
    };
    Ok(Value::from(n))
}

File: eval/lib/state_query.py
"""State queries against the test server (spec §4a).

12 query types — all HTTP GET. Ported faithfully from
crates/eval-runner/src/state_query.rs so cases behave identically under the
Python runner.
"""
from __future__ import annotations

import requests


def _unwrap_envelope(body):
    """Many NeoMind routes return {success, data}; return data if present."""
    if isinstance(body, dict) and "data" in body:
        return body["data"]
    return body


def _get_json(base: str, key: str, path: str):
    r = requests.get(f"{base}{path}", headers={"Authorization": f"Bearer {key}"}, timeout=30)
    body = r.json()
    return _unwrap_envelope(body)


def _exists(base: str, key: str, path: str) -> bool:
    r = requests.get(f"{base}{path}", headers={"Authorization": f"Bearer {key}"}, timeout=30)
    return r.status_code == 200


def _field(base: str, key: str, path: str, field: str):
    v = _get_json(base, key, path)
    if isinstance(v, dict):
        return v.get(field)
    return None


def _count(base: str, key: str, path: str) -> int:
    v = _get_json(base, key, path)
    if isinstance(v, list):
        return len(v)
    if isinstance(v, dict):
        for k in ("devices", "rules", "agents", "channels", "messages",
                  "dashboards", "automations", "data", "items"):
            arr = v.get(k)
            if isinstance(arr, list):
                return len(arr)
        for k in ("count", "total"):
            n = v.get(k)
            if isinstance(n, int):
                return n
    return 0


def _sid(params: dict, key: str) -> str:
    v = params.get(key)
    if not isinstance(v, str) or not v:
        raise ValueError(f"missing param {key}")
    return v


def _find_by_name(base: str, key: str, list_path: str, name: str) -> bool:
    """List collection + match by `name` field. Returns True on any match.

    Used by `*_exists` queries when the case specifies `name` instead of an
    explicit id — agents often create entities with auto-generated UUIDs that
    can't be predicted at case-authoring time.
    """
    v = _get_json(base, key, list_path)
    items: list = []
    if isinstance(v, list):
        items = v
    elif isinstance(v, dict):
        for k in ("devices", "rules", "agents", "channels", "messages",
                  "dashboards", "automations", "transforms", "data", "items"):
            arr = v.get(k)
            if isinstance(arr, list):
                items = arr
                break
    for it in items:
        if isinstance(it, dict) and it.get("name") == name:
            return True
    return False


def _id_or_name_exists(base: str, key: str, list_path: str,
                       get_path_template: str, params: dict) -> bool:
    """Existence check with transparent name fallback.

    Order:
    1. If `params.id` is given, try `GET /<resource>/{id}` directly.
       If 200, return True. If the value isn't a real UUID (e.g. a
       human-readable name like "rule-battery-low" that case-authors
       often pass as `id`), this GET returns 400/404 — fall through.
    2. Treat the value (id or name) as a display name and search the
       collection list for an entity whose `name` field matches.

    This mirrors how real users recover from "Invalid rule ID" errors and
    keeps existing case files working without modification.
    """
    val = params.get("id") or params.get("name")
    if not val:
        raise ValueError("state_query requires `id` or `name` param")
    if params.get("id"):
        if _exists(base, key, get_path_template.format(id=params["id"])):
            return True
    return _find_by_name(base, key, list_path, val)


def _resolve_id_or_name(base: str, key: str, list_path: str,
                        collection_key: str, params: dict) -> str | None:
    """Return an id suitable for GET /<resource>/{id}.

    Used by queries that need a full record (rule_enabled, agent_status,
    etc.). If `params.id` is a real UUID that resolves, return it as-is.
    Otherwise (no id, OR id is actually a name, OR id doesn't resolve),
    list the collection and find the entity whose `name` matches, then
    return its UUID.
    """
    val = params.get("id") or params.get("name")
    if not val:
        raise ValueError("state_query requires `id` or `name` param")
    if params.get("id"):
        # Probe: if a direct GET succeeds, the id is real.
        r = requests.get(
            f"{base}{list_path}/{params['id']}",
            headers={"Authorization": f"Bearer {key}"}, timeout=30,
        )
        if r.status_code == 200:
            return params["id"]
        # Else fall through and treat the value as a name.
    v = _get_json(base, key, list_path)
    items: list = []
    if isinstance(v, list):
        items = v
    elif isinstance(v, dict):
        arr = v.get(collection_key) if collection_key else None
        if isinstance(arr, list):
            items = arr
        else:
            for k in ("devices", "rules", "agents", "channels", "messages",
                      "dashboards", "automations", "transforms", "data", "items"):
                a = v.get(k)
                if isinstance(a, list):
                    items = a
                    break
    for it in items:
        if isinstance(it, dict) and it.get("name") == val:
            for id_key in ("id", "uuid", "automation_id", "rule_id", "agent_id"):
                if it.get(id_key):
                    return it[id_key]
    return None


def _get_with_name_fallback(base: str, key: str, list_path: str,
                            get_path_template: str, collection_key: str,
                            params: dict):
    """GET a single record by id (or by name→id resolution).

    Returns the unwrapped JSON dict (or {} if not found).
    """
    resolved = _resolve_id_or_name(base, key, list_path, collection_key, params)
    if not resolved:
        return {}
    v = _get_json(base, key, get_path_template.format(id=resolved))
    if isinstance(v, dict):
        # Many NeoMind endpoints wrap the record: {rule: {...}}, {agent: {...}}...
        for k in ("rule", "agent", "device", "dashboard", "automation", "channel"):
            inner = v.get(k)
            if isinstance(inner, dict):
                return inner
    return v if isinstance(v, dict) else {}


def _field_with_name_fallback(base: str, key: str, list_path: str,
                              get_path_template: str, collection_key: str,
                              params: dict, field: str):
    v = _get_with_name_fallback(base, key, list_path, get_path_template,
                                collection_key, params)
    return v.get(field) if isinstance(v, dict) else None


def run_query(q: dict, base: str, key: str) -> dict:
    """Run one state_query; returns {type, params, expected, actual, passed}.

    Supported assertion shapes:
    - `expected: <value>` → exact equality (default).
    - `expected_min: <int>` → numeric `actual >= expected_min`. Useful for
      "at least one message was sent" / "agent ran at least once".

    `*_exists` queries accept either `id` (direct lookup) or `name` (list
    + match by name). The name fallback exists because agents create
    entities with auto-generated UUIDs.
    """
    t = q["type"]
    params = q.get("params", {}) or {}
    expected = q.get("expected")
    expected_min = q.get("expected_min")

    if t == "device_exists":
        actual = _id_or_name_exists(base, key, "/devices", "/devices/{id}", params)
    elif t == "rule_exists":
        actual = _id_or_name_exists(base, key, "/rules", "/rules/{id}", params)
    elif t == "agent_exists":
        actual = _id_or_name_exists(base, key, "/agents", "/agents/{id}", params)
    elif t == "transform_exists":
        actual = _id_or_name_exists(base, key, "/automations", "/automations/{id}", params)
    elif t == "dashboard_exists":
        actual = _id_or_name_exists(base, key, "/dashboards", "/dashboards/{id}", params)
    elif t == "channel_exists":
        actual = _exists(base, key, f"/messages/channels/{_sid(params, 'name')}")
    elif t == "rule_enabled":
        actual = _field_with_name_fallback(base, key, "/rules", "/rules/{id}", "rules", params, "enabled")
    elif t == "agent_status":
        actual = _field_with_name_fallback(base, key, "/agents", "/agents/{id}", "agents", params, "status")
    elif t == "agent_execution_count":
        # GET /agents/:id returns stats.total_executions (line 2074 of agents.rs)
        v = _get_with_name_fallback(base, key, "/agents", "/agents/{id}", "agents", params)
        stats = v.get("stats") if isinstance(v, dict) else None
        actual = stats.get("total_executions") if isinstance(stats, dict) else 0
    elif t == "push_enabled":
        actual = _field(base, key, f"/data-push/{_sid(params, 'id')}", "enabled")
    elif t == "device_count":
        actual = _count(base, key, "/devices")
    elif t == "message_count":
        actual = _count(base, key, "/messages")
    elif t == "transform_count":
        # GET /automations returns all automations (transforms included);
        # the response carries either an `automations` array or a top-level list.
        actual = _count(base, key, "/automations")
    elif t == "dashboard_component_count":
        v = _get_with_name_fallback(base, key, "/dashboards", "/dashboards/{id}", "dashboards", params)
        comps = v.get("components") if isinstance(v, dict) else None
        actual = len(comps) if isinstance(comps, list) else 0
    else:
        raise ValueError(f"unknown state_query type: {t}")

    return {
        "type": t,
        "params": params,
        "expected": expected,
        "expected_min": expected_min,
        "actual": actual,
        "passed": (
            actual >= expected_min
            if expected_min is not None
            else actual == expected
        ),
    }
